vision/odlc_contour_filtering.py

```diff
@@ -0,0 +1,498 @@
+
+    Notes
+    -----
+    Possibly in future could modifiy the filter_contour() function to return two booleans
+    one to determine if there is a shape, another to determine if the detected shape was circular
+    (or an integer could be used, ie: 0=no shape, 1=shape, 2=circular shape)
+    """
+    # test hierarchy, bounding_box, min_area_box, and spikiness
+    if (
+        (not test_heirarchy(hierarchy, index))
+        or (not test_bounding_box(contours[index], image_dims))
+        or (not test_min_area_box(contours[index]))
+        or (not test_spikiness(contours[index]))
+    ):
+        return False
+    # then calculate approxPolyDP if above all pass (maybe change depending on test results)
+    if approx_contour is None:
+        approx_contour = cv2.approxPolyDP(
+            contours[index], cv2.arcLength(contours[index], True) * 0.05, True
+        )
+
+    # run polygon specific test
+    if not test_roughness(contours[index], approx_contour):
+        # if polygon test fails run circle test
+
+        # tuple[int, int, int, int] is the return type of cv2.boundingRect()
+        # the first two are the (y, x) of top left corner
+        # the last two are the (y, x) of the bottom right corner
+        cnt_bound_box_retval: tuple[int, int, int, int] = cv2.boundingRect(contours[index])
+        cnt_bound_box: bbox.BoundingBox = bbox.BoundingBox(
+            bbox.tlwh_to_vertices(
+                cnt_bound_box_retval[1],
+                cnt_bound_box_retval[0],
+                cnt_bound_box_retval[3] - cnt_bound_box_retval[1],
+                cnt_bound_box_retval[2] - cnt_bound_box_retval[0],
+            ),
+            bbox.ObjectType.STD_OBJECT,
+        )
+        # use _generate_mask to get a mask of the shape then cvt bool image to UInt8
+        cnt_mask: consts.Mask = generate_mask(contours[index], cnt_bound_box)
+        cnt_sc_img: consts.ScImage = np.where(cnt_mask, 255, 0).astype(np.uint8)
+
+        if not test_circleness(cnt_sc_img):
+            return False
+
+    return True
+
+
+# All of main is some basic testing code
+if __name__ == "__main__":
+    # create a blank image
+    test_image1: consts.Image = np.zeros([5000, 5000, 3], dtype=UInt8)
+
+    # define some points to make a polygon, there is some draw circle function to try circles also
+    raw_pts: NDArray[Shape["*, 2"], IntC] = np.array(
+        [[249, 0], [499, 249], [249, 499], [0, 249]], IntC
+    )
+    pts: consts.Contour = raw_pts.reshape((-1, 1, 2))
+    # put the points on the image
+    test_image1 = cv2.polylines(test_image1, [pts], True, (255, 255, 255))
+
+    cv2.imshow("pic", test_image1)
+    cv2.waitKey(0)
+    # make single channel to do contour stuff
+    test_image: consts.ScImage = cv2.cvtColor(test_image1, cv2.COLOR_BGR2GRAY)
+
+    # a variable length tuple is not good practice, but that is what opencv does here
+    # cnts_tmp: tuple(consts.Contour, ...) # also this gives a TypeError for some reason
```

Comment:
Is it giving an error because you are using parenthesis instead of brackets?
```suggestion
    # a variable length tuple is not good practice, but that is what opencv does here
    # cnts_tmp: tuple[consts.Contour, ...] # also this gives a TypeError for some reason
```